pkg/quarterdeck/invites.go

```diff
@@ -0,0 +1,203 @@
+package quarterdeck
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/gin-gonic/gin"
+	"github.com/oklog/ulid/v2"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/api/v1"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/db"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/db/models"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/middleware"
+	perms "github.com/rotationalio/ensign/pkg/quarterdeck/permissions"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/tokens"
+	"github.com/rotationalio/ensign/pkg/utils/sentry"
+	"github.com/rotationalio/ensign/pkg/utils/tasks"
+	"github.com/rotationalio/ensign/pkg/utils/ulids"
+)
+
+// InvitePreview returns details for a user invitation. This is a publicly accessible
+// endpoint because it has to be accessed by unauthenticated users without any claims.
+// Although the endpoint is read-only, it could return sensitive information about an
+// organization so this endpoint relies on the token being difficult to guess.
+func (s *Server) InvitePreview(c *gin.Context) {
+	var err error
+
+	// Get the token from the URL param
+	token := c.Param("token")
+
+	// Retrieve the invite from the database
+	var invite *models.UserInvitation
+	if invite, err = models.GetUserInvite(c.Request.Context(), token); err != nil {
+		if errors.Is(err, models.ErrNotFound) {
+			c.JSON(http.StatusBadRequest, api.ErrorResponse("invalid invitation"))
+			return
+		}
+
+		sentry.Error(c).Err(err).Msg("could not retrieve user invite")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not retrieve invitation"))
+		return
+	}
+
+	// Ensure the role is a recognized role
+	if !perms.IsRole(invite.Role) {
+		sentry.Warn(c).Str("role", invite.Role).Msg("invalid role for user invite")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("invalid invitation"))
+		return
+	}
```

Comment:
Any reason this check isn't in `invite.Validate`?

Reply:
Good point, I'll move it to that method.